src/openlcb/IfTcpImpl.hxx

```diff
@@ -0,0 +1,553 @@
+        if (current_time_ms > sequence_)
+        {
+            sequence_ = current_time_ms;
+        }
+        else
+        {
+            ++sequence_;
+        }
+        return sequence_;
+    }
+
+private:
+    /// Sequence number of last sent message.
+    long long sequence_ = 0;
+};
+
+/**
+ * This flow renders outgoing OpenLCB messages to their TCP stream
+ * representation.
+ */
+class TcpSendFlow : public MessageStateFlowBase
+{
+public:
+    TcpSendFlow(If *service, NodeID gateway_node_id,
+        HubPortInterface *send_target, HubPortInterface *skip_member,
+        SequenceNumberGenerator *sequence)
+        : MessageStateFlowBase(service)
+        , sendTarget_(send_target)
+        , skipMember_(skip_member)
+        , gatewayId_(gateway_node_id)
+        , sequenceNumberGenerator_(sequence)
+    {
+    }
+
+private:
+    Action entry() override
+    {
+        auto id = nmsg()->dst.id;
+        if (id)
+        {
+            auto *n = iface()->lookup_local_node(id);
+            if (n)
+            {
+                // Addressed message loopback.
+                auto *b = transfer_message();
+                b->data()->dstNode = n;
+                iface()->dispatcher()->send(b, priority());
+                return release_and_exit();
+            }
+        }
+        return allocate_and_call(sendTarget_, STATE(render_src_message));
+    }
+
+    Action render_src_message()
+    {
+        auto b = get_buffer_deleter(get_allocation_result(sendTarget_));
+        TcpDefs::render_tcp_message(*nmsg(), gatewayId_,
+            sequenceNumberGenerator_->get_sequence_number(), b->data());
+        b->data()->skipMember_ = skipMember_;
+        sendTarget_->send(b.release(), nmsg()->priority());
+
+        // Checks and performs global loopback.
+        if (!nmsg()->dst.id)
+        {
+            iface()->dispatcher()->send(transfer_message(), priority());
+        }
+        return release_and_exit();
+    }
+
+    /// @return the abstract message we are trying to send.
+    GenMessage *nmsg()
+    {
+        return message()->data();
+    }
+
+    /// @return owning interface object.
+    If *iface()
+    {
+        return static_cast<If *>(service());
+    }
+
+    /// Where to send the rendered messages to.
+    HubPortInterface *sendTarget_;
+    /// This value will be populated to the skipMember_ field.
+    HubPortInterface *skipMember_;
+    /// Populated into the source gateway field of the outgoing messages.
+    NodeID gatewayId_;
+    /// Responsible for generating the sequence numbers of the outgoing
+    /// messages.
+    SequenceNumberGenerator *sequenceNumberGenerator_;
+};
+
+/// This flow is listening to data from a TCP connection, segments the incoming
+/// data into TcpMessages based on the incoing size, and forwards packets
+/// containing the TCP message as string payload.
+class FdToTcpParser : public StateFlowBase
+{
+public:
+    FdToTcpParser(FdHubPortService *s, HubPortInterface *dst,
```

Comment:
Add method comment block here and state flow methods below.

Reply:
done